Assets/Scripts/Controllers/CharacterController.cs

```diff
@@ -23,6 +32,28 @@ private void FixedUpdate()
     }
 
 
+    private void Update()
+    {
+        // Check if the player has fired a weapon recently
+        if (_fireRate == 0)
+        {
+            if(Input.GetButton("Fire1"))
+            {
+                //Get values necessary to spawn projectiles outside of player collision
+                _projectileSpawnModifier = Input.mousePosition - _cam.WorldToScreenPoint(_rb.position);
+                _projectileSpawnModifier.Normalize();
+                _projectileSpawnModifier = _projectileSpawnModifier * _playerBoundingBox.size;
+
+                Instantiate(_weaponProjectile, _rb.position + _projectileSpawnModifier, Quaternion.identity);
+                _fireRate = 30;
+            }
+        }
+        else
+        {
+            _fireRate = _fireRate - 1;
```

Comment:
One thing that is important to consider is that the `_fireRate` is decremented _per-frame_, not on a time cycle.

This means that on my system which runs this scene at 600 fps, I'm blasting 20 projectiles a second. Someone with 30 fps will only fire once per second!

A better solution would be to determine the fire rate to be the **minimum number of seconds between two shots**. Key to doing so is using `[Time.time](https://docs.unity3d.com/ScriptReference/Time-time.html)`.

When the player shoots, mark down the time that they fired on (`_lastFireTime`). The next time they try to shoot on a subsequent frame, check and see if the new `Time.time` is greater than `_lastFireTime + _fireRate`. Essentually, we keep track of the next time the player **can** fire and disallow the user from shooting until we pass that time.

Let me know if you need any help with implementing this.